syntax = "proto3";
 
package clairvoyantedge;
 
message Segment{
    string segmentId = 1;
    double popularity = 2;
    int32 size = 3;
    int64 lastAccessed = 4;
    repeated string nodeIds = 5;
 }

message Video{
    string videoId = 1;
    repeated Segment segments = 2;
}

message Position{
    double latitude = 1;
    double longitude = 2;
} 

message NodeInfo{
    string nodeId = 1;
    Position position = 2;
}

message NodeSegments{
    string nodeId = 1;
    repeated string segmentIds = 2;
}

message Route{
    string routeId = 1;
    repeated NodeSegments placements= 2;
}


message GetVideoInfoRequest{
    string videoId = 1;
}

message GetVideoInfoResponse{
    Video video = 1;
}

message GetNearestNodesRequest{
    repeated Position positions = 1;
}

message GetNearestNodesResponse{
    repeated NodeInfo nodes = 1;
}

message SegmentOperation{
    string operation = 1;
    Segment segment = 2;
}

message UpdateSegmentInfoRequest{
    repeated SegmentOperation segmentOperations = 1;
}

message StatusResponse{
    repeated string statuses = 1;
}

message AddVideoInfoRequest{
    repeated Video videos = 1;
}

message AddNodeToRouteRequest{
    string routeId = 1;
    string nodeId = 2;
    repeated string segments = 3;
}

message DeleteRouteRequest{
    string routeId = 1;
}

service MetadataServer {
    rpc getVideoInfo (GetVideoInfoRequest) returns (GetVideoInfoResponse) {}
    rpc getNearestNodes (GetNearestNodesRequest) returns (GetNearestNodesResponse) {}
    rpc updateSegmentInfo (UpdateSegmentInfoRequest) returns (StatusResponse) {}
    rpc addVideoInfo (AddVideoInfoRequest) returns (StatusResponse) {}
    rpc addNodeToRoute (AddNodeToRouteRequest) returns (StatusResponse) {}
}
